### ISSUE TITLE
Encoders/Decoders don't get registered with cidlink.MulticodecDecoder unless you touch them

### ISSUE THREAD

**Issue by rvagg** [ipld/go-ipld-prime#40]
If you just want to consume via cidlink and not directly with dagcbor or dagjson, you don't get the benefit of `init()` in dagjson/multicodec.go or dagcbor/multicodec.go registering themselves for you. You have to invoke them with an awkward construct like `_ = dagcbor.Encoder` (or maybe there's a better way?). Perhaps a public `dagcbor.Register()` or something simple would suffice? Or would it hurt so much to have them registered by default in cidlink rather than the other way around?

**Comment by warpfork** [comment]
Yeah, I... don't really like this either.  I'm surprised I even wrote it this way -- I dislike underscore-imports as an action mechanism whenever I run across them (e.g., in the standard library, "image/png" and friends have some behaviors like this).  So I'd be broadly very open to considering different mechanisms for this.

But, setting aside my feelings of squick for a moment, there are a few things this design has going for it:

- transitive import minimization -- granularly, you don't get the encoding packages if you don't use them.
- minimum complexity at use site -- admittedly, at the cost of using a global variable -- but surely "surely" in this case the global is fine, since it's not sane to have different multicodec ints in different parts of a program.
- doing things in the package `init` is race-free -- package init happens once, and it happens before user code starts, so it's one of the easiest ways to initialize a shared map that's race-free.

---

To specific questions you raised:

> You have to invoke them with an awkward construct like _ = dagcbor.Encoder (or maybe there's a better way?).

There is another way (though not seen very often):

```go
import (
    _ "package.example.org/foo" // underscore imports don't cause complaints, and trigger init
)
```

It's the eye of the beholder whether this is "better", perhaps...
> Perhaps a public dagcbor.Register() or something simple would suffice?


See third point above -- we could do this if made us feel better, but arguably, it should just then be a dummy method anyway, because the importing is the effect.

> Or would it hurt so much to have them registered by default in cidlink rather than the other way around?

---

The biggest thing I don't like about side-effecting imports is that they mean as long as _some_ part of your program imports something that (possibly transitively) imports those packages... your program works.  And if you don't, it becomes silently wrong.  The maintainability properties of this are IMO not great.

But this is also true with the idea of a slightly-more-explicit `Register` method.  It's really the global map that creates this issue.

**Comment by warpfork** [comment]
We could make a big change to get rid of the global map entirely by making the multicodec registry into something attached to a `Context` and passed down functionally thusly.  I'm not sure if the result would be pleasingly terse, though.

**Comment by rvagg** [comment]
This seems like a question of idiomatic Go which I'm far from qualified to give answers to. I'll use whatever I need to make it work but I'd like to know that I'm doing _normal_ Go that a _normal_ Go programmer won't wince at. Also, docs are going to be helpful. The tests aren't helpful in this cause because they're very <i>unit</i>y rather than <i>integration</i>y, I had to stumble through this one.

I'm happy with the `import _ "package.example.org/foo` approach for now, that seems reasonable to my unidiomatic mind.

**Comment by hannahhoward** [comment]
so I just encountered this one and remembered it's a thing.

one potential take is the moment I have `cidlink`, I have `go-cid` and I've already bit the big transitive dependency bullet, and therefore maybe registering a couple encoders by default is not a bad thing.

BUT I think a better starting place is to look at this along with NodeBuilderChooser as a whole, and take another shot at a design, at some point, because that has come up as controversial as well.
